libs/helpers/helpers.go

```diff
@@ -0,0 +1,81 @@
+package helpers
+
+import (
+	"log"
+	"strconv"
+	"strings"
+)
+
+// LatestVersion gets a slice of versions and returns the latest version as a string
+// e.g. ["0.1", "0.10", "1.0"] or ["0.1.0", "0.10.1", "1.0.2"]
+// any number of minor/patch versions will work as long as all versions are consistent
+// i.e. this is not allowed ["0.1", "0.10.0", "1.0"] and the function will return ""
+func LatestVersion(vl []string) string {
+	if len(vl) == 0 {
+		return ""
+	}
+
+	var ss [][]string
+	var vlen int
+
+	// split all versions into their atomic parts (e.g. "1.0" becomes ["1", "0"]) and put into slice of slices ss
+	for _, v := range vl {
+		sv := strings.Split(v, ".")
+		if vlen == 0 {
+			vlen = len(sv)
+		} else {
+			if vlen != len(sv) {
+				log.Printf("Error: format mismatch for version '%s'", v)
+				return ""
+			}
+		}
+		ss = append(ss, sv)
+	}
+
+	// bubble sort the slices starting with the minor version (the one on the right) and moving to the left
+	for l := vlen - 1; l >= 0; l-- {
+		for m := 0; m < len(ss); m++ {
+			for n := 0; n < len(ss)-1; n++ {
+				int1, err := strconv.Atoi(ss[n][l])
+				if err != nil {
+					log.Printf("Error: format mismatch for version '%s'", strings.Join(ss[n], "."))
+					return ""
+				}
+				int2, err := strconv.Atoi(ss[n+1][l])
+				if err != nil {
+					log.Printf("Error: format mismatch for version '%s'", strings.Join(ss[n+1], "."))
+					return ""
+				}
+				if int1 > int2 {
+					ss[n], ss[n+1] = ss[n+1], ss[n]
+				}
+			}
+		}
+	}
+
+	return strings.Join(ss[len(ss)-1], ".")
+}
```

Comment:
I didn't make it take arbitrary length strings, although that should be easy.... this is quite a bit faster

```
➜  tmp go test -bench=. -benchtime=20s
goos: darwin
goarch: amd64
BenchmarkLatestVersionT-8   	10000000	      3858 ns/op
BenchmarkLatestVersionC-8   	10000000	      2741 ns/op
BenchmarkLatestVersionN-8   	10000000	      2436 ns/op
```

```golang
func LVS(vl []string) string {
	if len(vl) == 0 {
		return ""
	}

	// initialize latest with zeros
	latest := []int{}
	for i := 0; i < len(strings.Split(vl[0], ".")); i++ {
		latest = append(latest, 0)
	}

	// loop over list and compare each to latest
	for _, sv := range vl {
		current := strings.Split(sv, ".")
		if len(current) == 0 {
			log.Printf("Error: bad current version")
			return ""
		}

		si := []int{}
		for _, s := range current {
			i, err := strconv.Atoi(s)
			if err != nil {
				log.Println("Failed to parse", current)
				return ""
			}
			si = append(si, i)
		}

		if len(latest) != len(si) {
			log.Println("Error format mismatch for", sv)
			return ""
		}

		for i, siv := range si {
			if siv > latest[i] {
				latest = si
				break
			}
		}
	}

	l := []byte{}
	for _, i := range latest {
		l = strconv.AppendInt(l, int64(i), 10)
	}
	return string(l) 
}
```

Reply:
hmm, yeah, no need to sort all the versions if we only need to pick the latest one